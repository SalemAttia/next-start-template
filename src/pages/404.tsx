import Error from 'next/error'

export default function NotFound() {
  return <div>
      i am not here
  </div>
}