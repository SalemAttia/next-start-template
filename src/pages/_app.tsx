import axios from 'axios'
import Head from 'next/head'
import React, { useEffect } from 'react'
import { SWRConfig } from 'swr'
import * as Sentry from '@sentry/node'
import * as gtag from '../../scripts/gtag'
import { Router } from 'next/dist/client/router'

if (process.env.NEXT_PUBLIC_SENTRY_DSN) {
  Sentry.init({
    enabled: process.env.NODE_ENV === 'production',
    dsn: process.env.NEXT_PUBLIC_SENTRY_DSN,
  })
}

const App = ({ Component, pageProps }) => {
  const isProduction = process.env.NODE_ENV === 'production'
  // google analitc configration
  if (isProduction) {
    useEffect(() => {
      const handleRouteChange = (url) => {
        gtag.pageview(url)
      }
      Router.events.on('routeChangeComplete', handleRouteChange)
      return () => {
        Router.events.off('routeChangeComplete', handleRouteChange)
      }
    }, [])
  }

  return (
    <>
      <Head>
        <title>Template main title</title>
        <meta
          name="viewport"
          content="minimum-scale=1, initial-scale=1, width=device-width"
        />
      </Head>
      <SWRConfig
        value={{ fetcher: (url: string) => axios(url).then((r) => r.data) }}
      >
        <div>
          <Component {...pageProps} />
        </div>
      </SWRConfig>
    </>
  )
}

export default App
