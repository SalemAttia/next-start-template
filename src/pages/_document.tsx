import Document, { Html, Head, Main, NextScript, DocumentContext } from 'next/document'

class MyDocument extends Document {
  static async getInitialProps(ctx: DocumentContext) {
    const originalRenderPage = ctx.renderPage

    ctx.renderPage = () =>
      originalRenderPage({
        // useful for wrapping the whole react tree
        enhanceApp: (App) => App,
        // useful for wrapping in a per-page basis
        enhanceComponent: (Component) => Component,
      })

    // Run the parent `getInitialProps`, it now includes the custom `renderPage`
    const initialProps = await Document.getInitialProps(ctx)

    return initialProps
  }

  render() {
    return (
      <Html>
        <Head />
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument