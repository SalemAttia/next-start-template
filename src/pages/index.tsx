import Head from 'next/head'
import HomeComponent from '../components/home'
import Layout from '../components/layout'

export const Home = (): JSX.Element => (
  <Layout>
    <Head>
      <title>Create Next App</title>
      <link rel="icon" href="/favicon.ico" />
    </Head>
    <HomeComponent />
  </Layout>
)

export default Home
