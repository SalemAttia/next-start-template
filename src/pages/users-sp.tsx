import { GetStaticProps } from 'next'
import { UserModel } from '../api/User'
import { getUsers } from '../lib/api'

interface UserProps {
  users: UserModel[]
}

export default function User({ users }: UserProps) {
  return (
    <div>
      {users.map((f) => (
        <div key={f.id}>
          <ul>
            <li>{f.name}</li>
            <li>{f.email}</li>
            <li>{f.username}</li>
          </ul>
        </div>
      ))}
    </div>
  )
}

export const getStaticProps: GetStaticProps<UserProps> = async () => {
  // build time render only
  const users = ((await getUsers()) as UserModel[]) || []
  return { props: { users } }
}
