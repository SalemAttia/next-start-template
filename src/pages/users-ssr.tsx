import { GetServerSideProps } from 'next'
import { UserModel } from '../api/User'
import { getUsers } from '../lib/api'

interface UserProps {
  users: UserModel[]
}

export const User = ({ users }: UserProps): JSX.Element => {
  return (
    <div className="center">
      {users.map((f) => (
        <div key={f.id}>
          <ul>
            <li>{f.name}</li>
            <li>{f.email}</li>
            <li>{f.username}</li>
          </ul>
        </div>
      ))}
    </div>
  )
}
export default User

export const getServerSideProps: GetServerSideProps<UserProps> = async () => {
  // call the server each request pre render and it block the page until it gets load
  const users = ((await getUsers()) as UserModel[]) || []
  return { props: { users } }
}
