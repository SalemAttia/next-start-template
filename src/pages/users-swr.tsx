import useSWR from 'swr'
import { UserModel } from '../api/User'

interface UserProps {
  users: UserModel[]
}

export default function user() {
  // loding the api after the page is loaded
  const { data } = (useSWR(
    'https://jsonplaceholder.typicode.com/users'
  ) as unknown) as { data: UserModel[] }

  return (
    <div>
      <h2>HERE</h2>
      {data ? (
        data.map((f) => (
          <div key={f.id}>
            <ul>
              <li>{f.name}</li>
              <li>{f.email}</li>
              <li>{f.username}</li>
            </ul>
          </div>
        ))
      ) : (
        <div>loding...</div>
      )}
    </div>
  )
}
