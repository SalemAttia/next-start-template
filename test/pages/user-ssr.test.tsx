import React from 'react'
import { render } from '../testUtils'
import { User } from '../../src/pages/users-ssr'

describe('User ssr page', () => {
  it('matches snapshot', () => {
    const data = []
    const { asFragment } = render(<User users={data} />, {})
    expect(asFragment()).toMatchSnapshot()
  })
})
