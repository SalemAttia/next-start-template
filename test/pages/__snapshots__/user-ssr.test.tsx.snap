// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User ssr page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="center"
  />
</DocumentFragment>
`;
